core/embed/rust/src/ui/display/mod.rs

```diff
@@ -325,17 +227,14 @@ pub fn rect_rounded2_partial(
     let mut icon_data = [0_u8; ((MAX_ICON_SIZE * MAX_ICON_SIZE) / 2) as usize];
     let mut icon_width = 0;
 
-    if let Some((icon_bytes, icon_color)) = icon {
-        let (toif_size, toif_data) = toif_info_ensure(icon_bytes, ToifFormat::GrayScaleEH);
-
-        if toif_size.x <= MAX_ICON_SIZE && toif_size.y <= MAX_ICON_SIZE {
-            icon_area = Rect::from_center_and_size(center, toif_size);
+    if let Some((icon, icon_color)) = icon {
+        if icon.width() <= MAX_ICON_SIZE && icon.height() <= MAX_ICON_SIZE {
+            icon_area = Rect::from_center_and_size(center, icon.size());
             icon_area_clamped = icon_area.clamp(constant::screen());
-
-            let mut ctx = UzlibContext::new(toif_data, None);
+            let mut ctx = UzlibContext::new(icon.data(), None);
             unwrap!(ctx.uncompress(&mut icon_data), "Decompression failed");
```

Comment:
this could be a `fn uncompress(&self, buffer)` on `Toif`.
(and have it just straight out panic if it fails, I see no value in passing out an error only to have to unwrap! it one level up)

might actually be a tiny bit nicer to do `fn decompression_context(&self, window) -> UzlibContext` on Toif too, although the only thing that's achieving is hiding the raw `data`

Reply:
https://github.com/trezor/trezor-firmware/commit/bc174f7ed04e27f751b67a4e52dd4985ae773c12